src/main/java/de/dennisguse/opentracks/viewmodels/IntervalStatistics.java

```diff
@@ -9,44 +10,64 @@
 
 public class IntervalStatistics {
-    private final List<Interval> intervalList = new ArrayList<>();
+    private TrackStatisticsUpdater trackStatisticsUpdater = new TrackStatisticsUpdater();
+    private final List<Interval> intervalList;
     private final Distance distanceInterval;
+    private final Distance minGPSDistance;
+    private Interval interval, lastInterval;
 
-    public IntervalStatistics(@NonNull List<TrackPoint> trackPoints, Distance distanceInterval) {
-        intervalList.clear();
+    /**
+     * @param distanceInterval distance of every interval.
+     * @param minGPSDistance   the setting value for GPS distance.
+     */
+    public IntervalStatistics(Distance distanceInterval, Distance minGPSDistance) {
         this.distanceInterval = distanceInterval;
+        this.minGPSDistance = minGPSDistance;
 
+        interval = new Interval();
+        lastInterval = new Interval();
+        intervalList = new ArrayList<>();
+        intervalList.add(lastInterval);
+    }
+
+    public void addTrackPoints(@NonNull List<TrackPoint> trackPoints) {
         if (trackPoints.size() == 0) {
             return;
         }
 
-        Interval interval = new Interval();
-        interval.gain_m += trackPoints.get(0).hasAltitudeGain() ? trackPoints.get(0).getAltitudeGain() : 0;
-        interval.loss_m += trackPoints.get(0).hasAltitudeLoss() ? trackPoints.get(0).getAltitudeLoss() : 0;
-        for (int i = 1; i < trackPoints.size(); i++) {
-            TrackPoint prevTrackPoint = trackPoints.get(i - 1);
-            TrackPoint trackPoint = trackPoints.get(i);
+        boolean newIntervalAdded = false;
+        for (TrackPoint trackPoint : trackPoints) {
+            trackStatisticsUpdater.addTrackPoint(trackPoint, minGPSDistance);
+
+            if (trackStatisticsUpdater.getTrackStatistics().getTotalDistance().plus(interval.distance).greaterOrEqualThan(distanceInterval)) {
+                interval.update(trackStatisticsUpdater.getTrackStatistics(), trackPoint);
 
-            if (trackPoint.hasLocation() && prevTrackPoint.hasLocation()) {
-                interval.distance = interval.distance.plus(prevTrackPoint.distanceToPrevious(trackPoint));
-                interval.time = interval.time.plus(Duration.between(prevTrackPoint.getTime(), trackPoint.getTime()));
-                interval.gain_m += trackPoint.hasAltitudeGain() ? trackPoint.getAltitudeGain() : 0;
-                interval.loss_m += trackPoint.hasAltitudeLoss() ? trackPoint.getAltitudeLoss() : 0;
+                double adjustFactor = distanceInterval.dividedBy(interval.distance);
+                Interval adjustedInterval = new Interval(interval);
+                adjustedInterval.adjust(adjustFactor);
 
-                if (interval.distance.greaterThan(distanceInterval)) {
-                    Interval adjustedInterval = new Interval(interval, distanceInterval.dividedBy(interval.distance));
+                intervalList.remove(intervalList.size() - 1);
```

Comment:
this is replacing the last element?
Then `intervalList.set(intervalList.size() - 1, adjustedInterval)` can be used.